File: LeetCode - Algorithm/Binary Tree/114.java
// 114. Flatten Binary Tree to Linked List
// https://leetcode.com/problems/flatten-binary-tree-to-linked-list/

import java.util.Stack;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public void flatten(TreeNode root) {
        if (root == null)
            return;

        Stack<TreeNode> stack = new Stack<>();

        stack.add(root);

        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();

            if (node.right != null)
                stack.add(node.right);
            if (node.left != null)
                stack.add(node.left);

            node.left = null;
            if (!stack.isEmpty())
                node.right = stack.peek();
        }
    }
}

// [5, 4, 3]
// 2

File: LeetCode - Algorithm/Binary Tree/114.js
// 114. Flatten Binary Tree to Linked List
// https://leetcode.com/problems/flatten-binary-tree-to-linked-list/

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function (root) {
  getEnd(root);
};

const getEnd = (node) => {
  if (!node) return;

  if (node.left && node.right) {
    const leftEnd = getEnd(node.left);
    const rightEnd = getEnd(node.right);

    const temp = node.right;
    node.right = node.left;
    node.left = null;
    leftEnd.right = temp;

    return rightEnd;
  } else if (!node.left && node.right) {
    const rightEnd = getEnd(node.right);

    return rightEnd;
  } else if (node.left && !node.right) {
    const leftEnd = getEnd(node.left);

    node.right = node.left;
    node.left = null;

    return leftEnd;
  } else {
    return node;
  }
};
